source/mir/random/nonuniform.d

```diff
@@ -0,0 +1,246 @@
+*/
+auto exponential(T, R = uint)()
+{
+    import mir.internal.math : exp, log;
+
+    auto pdf    = (T x) => cast(T) exp(-x);
+    auto invPdf = (T x) => cast(T) -log(x);
+
+    // values from [Marsaglia00]
+    enum fallback = q{
+        T fallback(RNG)(ref RNG gen) const
+        {
+            import std.random : uniform;
+            auto u = uniform!("[]", T, T)(0, 1, gen);
+            return 7.69711 - u;
+        }
+    };
+
+    return Ziggurat!(T, fallback, R, false)(pdf, invPdf, 256, T(7.697117470131487), T(3.949659822581572e-3));
+}
+
+/**
+Ziggurat algorithm for generating a random variable from a montone, decreasing density.
+
+Complexity: O_avg: O(1)
+
+References:
+    Marsaglia, George, and Wai Wan Tsang. "The ziggurat method for generating random variables."
+    Journal of statistical software 5.8 (2000): 1-7.
+*/
+struct Ziggurat(T, string _fallback, R = uint, bool bothSides)
+    if (isNumeric!T)
+{
+
+private:
+
+    /// monotone decreasing probability density function
+    /// constants don't matter
+    T function(T x) pdf;
+    T function(T x) invPdf;
+
+    /// precalculate difference x_i / x_{i+1}
+    T[] xDiv;
+    /// precalculate scaling to R.max for x_i
+    T[] xScaled;
+    /// precalculate pdf value for x_i
+    T[] fs;
+
+    /// number of blocks
+    size_t k;
+    // mask to use to get the first log_2 k bits (=k - 1)
+    size_t kMask;
+
+    /// left-point of the right-most block
+    T rightEnd;
+
+    /// area of every column
+    T blockArea;
+
+public:
+    this(T function(T x) pdf, T function(T x) invPdf, size_t k, T rightEnd,
+         T blockArea)
+    {
+        this.pdf = pdf;
+        this.invPdf = invPdf;
+        this.k = k;
+        this.rightEnd = rightEnd;
+        this.blockArea = blockArea;
+
+        // scale factor to the range of UIntType
+        T maxT = bothSides ? R.max / 2 : R.max;
+        kMask = k - 1;
+
+        auto xn = rightEnd; // Next x_{i+1}
+        auto xc = xn; // Current x_i
+        auto fn = pdf(xn);
+
+        // k_i = floor(2^32 (x_{i-1} / x_i))
+        xDiv    = new T[k];
+        xDiv[0] = (xn * fn / blockArea) * maxT;
+        xDiv[1] = 0;
+
+        // w_i = 0.5^32 * x_i
+        xScaled        = new T[k];
+        xScaled[0]     = (blockArea / fn) / maxT;
+        xScaled[$ - 1] = xn / maxT;
+
+        // f_i = f(x_i)
+        fs        = new T[k];
+        fs[0]     = T(1);
+        fs[$ - 1] = fn;
+
+        for (auto i = k - 2; i >= 1; i--)
+        {
+            xn = invPdf(blockArea / xn + fn);
+            xDiv[i + 1] = (xn / xc) * maxT;
+            fn = fs[i] = pdf(xn);
+            xScaled[i] = xn / maxT;
+            xc = xn;
+        }
+    }
+
+    /// samples a value from the discrete distribution
+    T opCall() const
+    {
+        import std.random : rndGen;
+        return opCall(rndGen);
+    }
+
+    /// samples a value from the discrete distribution using a custom random generator
+    T opCall(RNG)(ref RNG gen) const
```

Comment:
> Assuming you do need to know the word-type (your template parameter R) at compile time, you might want to validate that typeof(gen.front) matches it.

Nice idea (done).

> Note that in principle at least it's possible to handle different R and generator return type: if R's size is a multiple of the generator's word size, then you can populate it by several calls to the generator; conversely, if R is smaller than the generator's word size, you can use a single RNG variate to provide several of the needed values. But this may be adding excessive complexity.

Good point - how common is it to have something different than `uint` or `ulong`?